jax/_src/nn/functions.py

```diff
@@ -844,24 +871,29 @@ def dot_product_attention(
   :code:`probs` as the output of :math:`softmax`.
 
   Throughout this function, we utilize the following uppercase letters to
-  represent the shape of array:
+  represent the shape of array::
 
     B = batch size
     S = length of the key/value (source)
     T = length of the query (target)
     N = number of attention heads
     H = dimensions of each attention head
+    K = number of key/value heads
+    G = number of groups, which equals to N // K
 
   Args:
     query: query array; shape :code:`(BTNH)`
-    key: key array; shape :code:`(BSNH)`
-    value: value array; shape :code:`(BSNH)`
-    bias: optional, bias array to be added to logits; shape broadcastable to
-      :code:`(BNTS)`.
+    key: key array: shape :code:`(BSKH)`. When `K` equals `N`, multi-headed
+      attention (MHA) is performed. Otherwise, when `N` is a multiple of `K`,
+      grouped query attention (GQA) is performed. Additionally, it supports the
+      shape :code:`(BSH)` for multi-query attention (MQA).
```

Comment:
Why do we need to support a different rank here? Isn't MQA just K=1? (ie group size 1) Complicates a lot of code having to check for this, and will make jaxtyping type signatures more complicated if we ever want to support them.

Reply:
Sure. Removed.